### PR TITLE
[stable9.1] make JobList::next() lock free

### PR DESCRIPTION
**Pull request by butonic**
## Description

adds the reserved_at and last_checked columns to the wuere clause of the update. This allows us to check if the row was updated by this execution and call next() in case another thread already updated this job. 

## Related Issue

fixes https://github.com/owncloud/core/issues/27071

## Motivation and Context

locking issues when running cron from multiple machines on galera cluster

## How Has This Been Tested?

cli, but needs jenkins and verification by @cdamken first, maybe also @Helios07 can test this works as intended in larger installations. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by PVince81** [comment]
adding @jvillafanez, I'm pretty sure he'll find something 😉

**Comment by butonic** [comment]
cc @felixboehm because he may know admins that may help test this patch in the wild

File: lib/private/BackgroundJob/JobList.php [hunk 1]
```diff
@@ -186,18 +186,24 @@ public function getNext() {
...
 			$update->setParameter('jobid', $row['id']);
-			$update->execute();
-			$this->connection->unlockTable();
+			$update->setParameter('reserved_at', $row['reserved_at']);
+			$update->setParameter('last_checked', $row['last_checked']);
+			$count = $update->execute();
 
+			if ($count === 0) {
+				// Background job already executed elsewhere, try again.
+				return $this->getNext();
```
Comment (PVince81) [inline lib/private/BackgroundJob/JobList.php]: Recursion sounds a bit dangerous in the event where a lot of consecutive jobs are discovered that were already executed elsewhere. How realistic is this scenario ?

**Comment by PVince81** [comment]
`getTime()` returns a value in seconds.

There is a slight risk that a concurrent process will set the timestamp to the same value as before if running within less than a second, and this process would think it was the same value as before when it wasn't.

@butonic can you confirm whether this is a potential issue ?

That could be the reason why lockTable() was used instead of this approach

A possible solution would be to increase the resolution of timestamp, but that would only reduce the risk to milliseconds, not eliminate them.

Should we also add a "update_counter"  value that increases every time the value is changed ?
So if a concurrent process changed the reserved_at value even to the same value, the counter would always be increased.
So we only update if the "update_counter" has the same value as for the select.

@butonic thoughts ?

**Comment by PVince81** [comment]
The more time passes, the more I feel convinced about using the "update_counter" approach instead which should eliminate any timing issues.

Any drawbacks against that ?

**Comment by butonic** [comment]
The UPDATE is atomic, which is why no two concurrent processes can update the same job. The first will change the timestamps, the second cannot update the same row because it no longer finds the matching row.

**Comment by PVince81** [comment]
I re-read the query, and it seems the problem I thought seeing doesn't exist, because the select picks any job olders than 12 hours. I only saw the "lte" and thought it could select a job that had a timestamp that was just updated within the same second.

All fine then, looks good. Needs testing on real env.

**Comment by PVince81** [comment]
:+1:

**Comment by PVince81** [comment]
Port to master: https://github.com/owncloud/core/pull/27599

**Comment by PVince81** [comment]
Did some local testing manually running cron.php, works fine.

**Pull request merged**